CommonTools/PileupAlgos/plugins/PuppiProducer.cc

```diff
@@ -283,58 +290,81 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
                           lWeights[iCand] * aCand.pz(),
                           lWeights[iCand] * aCand.energy());
 
+    // Here, we are using existing weights, or we're using packed candidates.
+    // That is, whether or not we recomputed the weights, we store the
+    // source candidate appropriately, and set the p4 of the packed candidate.
     if (fUseExistingWeights || fClonePackedCands) {
       pCand->setP4(puppiP4s.back());
       pCand->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
-      fPackedPuppiCandidates->push_back(*pCand);
+      fPackedPuppiCandidates.push_back(*pCand);
     } else {
-      pfCand->setP4(puppiP4s.back());
+      // Here, we are not using packed candidates. That is, this is MINIAOD
+      // and we are computing puppi for the PFCandidates themselves.
+      // We have TWO collections to write here. The first is the "weighted"
+      // version that weights the p4 by the puppi weight. The second
+      // is the standard version, where we just set the internal weight
+      // and do not adjust the p4. Only the second is persistified in
+      // AOD. The other can be used by user modules to have their own
+      // PFCandidate collection, and as a check for the weighting procedure.
+      pfCandWeighted->setP4(puppiP4s.back());
+      pfCandWeighted->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
+      pfCand->setP4(aCand.p4());
       pfCand->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
-      fPuppiCandidates->push_back(*pfCand);
+      if (fPuppiForLeptons) {
+        pfCandWeighted->setPuppiWeight(1.0, lWeights[iCand]);
+        pfCand->setPuppiWeight(1.0, lWeights[iCand]);
+      } else {
+        pfCandWeighted->setPuppiWeight(lWeights[iCand]);
+        pfCand->setPuppiWeight(lWeights[iCand]);
+      }
```

Comment:
Shouldn't there be a way to set both weights here (by reading a collection with the other weight already set)?

is there a way to compute both weights in one go?

Reply:
Not how it's currently set up. The weights are computed with the same code upstream. If we want them to be run together I have to rewrite the module.